### DISCUSSION TITLE
How i can update token, when i do request in page, on server side. NextJS.

### DISCUSSION THREAD

**Discussion by stanislavstru** [vvo/iron-session#727] [Q&A] [answered]
Please help me. I trying set new token after refresh token. 

I have this component. 

<img width="840" alt="image" src="https://github.com/vvo/iron-session/assets/106296114/973d854a-9df2-4b54-8c24-0461cfccf333">
 
on server side. 

I want add initial request and get initial date, but if this request begin update token, he can't save it.

Guys help me. How can i set new token, on this step? before client side part

**Comment by vvo** [comment]
Hey @stanislavstru did you ever fix your issue? At the moment your example is too specific for me to give you some help. Please try to re-explain what you're trying to do if you have some time.

**Comment by vvo** [comment, marked as answer]
@stanislavstru A server component cannot write cookies. That is a Next.js rule rather than an iron-session one: `cookies().set()` only works in a Server Action, a Route Handler, or `proxy.ts`. So a `session.save()` during a page render has nowhere to go.

To refresh a token before the page renders, do it in `proxy.ts` (`middleware.ts` before Next 16). v9 adds an adapter for it, and the value you save is visible to the render that follows:

```ts
import { NextResponse, type NextRequest } from "next/server";
import { getIronSession, nextProxyCookies } from "iron-session";

export async function proxy(request: NextRequest) {
  const response = NextResponse.next();
  const session = await getIronSession(nextProxyCookies(request, response), options);

  if (session.expiresAt && session.expiresAt < Date.now()) {
    session.token = await refreshToken(session.refreshToken);
    session.expiresAt = Date.now() + 3600 * 1000;
    await session.save();
  }

  return response;
}
```

Before v9 this did not work: the save looked like it succeeded and the cookie never arrived.

v9 notes: https://github.com/vvo/iron-session/releases/tag/v9.0.0
